src/component/RoadmapCanvas.js

```diff
@@ -3,101 +3,69 @@ import ContextMenu from "d3-v4-contextmenu";
     }
   }
 
-  _render(items, roadmap) {
-    const barHeight = 20;
-    const gap = barHeight + 4;
-    const topPadding = 20;
-
-    const h = items.length * gap + 40;
+  _render(items, groups, roadmap) {
     const w = this.targetElement.node().clientWidth;
+    const h = items.length * this.style.gap + 40;
 
-    let svg = this.targetElement.append("svg").attr("width", w).attr("style", "overflow: visible");
-    svg.attr("height", function() {
-      return parseInt(svg.attr("height") || 0, 10) + h;
-    });
+    let svg = this._generateSVG(w, h);
 
-    svg.on('contextmenu', function() {
-      d3.event.preventDefault();
-      let contextMenu = new ContextMenu([
-        {
-          label: "copy json data to clip board",
-          cb: function (e) {
-            let dummy = document.createElement("input");
-            document.body.appendChild(dummy);
-            dummy.setAttribute("id", "copy-dummy");
-            document.getElementById("copy-dummy").value = roadmap.toString();
-            dummy.select();
-            document.execCommand("copy");
-            document.body.removeChild(dummy);
-          }
-        }
-      ]);
-      contextMenu.show(svg, d3.mouse(this)[0], d3.mouse(this)[1]);
-    });
+    this._drawVerticalGroupBox(svg, groups, this.style.gap, this.style.topPadding, w);
 
-    let groups = [];
-    let total = 0;
-    for (let i = 0; i < items.length; i++){
-      let j = 0;
-      let found = false;
-      while (j < groups.length && !found) {
-        found = (groups[j].name === items[i].group);
-        j++;
-      }
-      if (!found) {
-        let count = 0;
-        j = 0;
-        while (j < items.length) {
-          if (items[j].group === items[i].group) {
-            count++;
-          }
-          j++;
-        }
-        groups.push({
-          type: "group",
-          name: items[i].group,
-          count: count,
-          previous: total,
-          style: items[i].style
-        });
-        total += count;
-      }
-    }
+    let labels = this._drawVerticalLabels(svg, groups, this.style.gap, this.style.topPadding);
 
-    items.sort(function(a, b) {
-      if (a.group === b.group) {
-        if (a.taskOrder !== b.taskOrder) {
-          return a.taskOrder > b.taskOrder ? 1 : -1;
-        }
-        return a.from > b.from ? 1 : -1;
-      } else {
-        if (a.order !== b.order) {
-          return a.order > b.order ? 1 : -1;
-        }
-        return a.group > b.group ? 1 : -1;
-      }
+    let sidePadding = labels.node().parentNode.getBBox().width + 15;
+
+    let xScale = this._generateXScale(items, w, sidePadding);
+    this._drwanXAxis(svg, xScale, sidePadding, this.style.topPadding);
+
+    this._drawItemLines(items, svg, xScale, this.style.gap, sidePadding, this.style.topPadding, this.style.barHeight);
+
+    this._addMouseHelper(svg, xScale, this.style.barHeight, sidePadding, roadmap);
+  }
+
+  _generateSVG(w, h) {
```

Comment:
Should do only `generate`, Current implmentation has 2 feature `generate` and `append`.